Engine/media/audio/sound.cpp

```diff
@@ -53,31 +56,153 @@ static int GuessSoundTypeFromExt(const String &extension)
     return 0;
 }
 
+// Sound cache, stores most recent used sounds, tracks use history with MRU list.
+// TODO: refactor into the resource cache class, share with the sprite cache.
+class SoundCache
+{
+public:
+    typedef std::shared_ptr<std::vector<uint8_t>> DataRef;
+
+    void SetMaxCacheSize(size_t size)
+    {
+        FreeMem(size);
+        _maxSize = size;
+    }
+
+    DataRef Get(const String &name)
+    {
+        const auto found = _map.find(name);
+        if (found == _map.end())
+            return DataRef();
+        // Move to the beginning of the MRU list
+        _mru.splice(_mru.begin(), _mru, found->second.MruIt);
+        return found->second.Data;
+    }
+
+    // Add particular item into the cache, remove oldest items if cache size is exceeded
+    void Put(const String &name, DataRef &ref)
+    {
+        assert(ref);
+        if (!ref)
+            return; // safety precaution
+        if (Get(name))
+            return; // already in cache
+        if (_maxSize == 0)
+            return; // cache is disabled
+        // Clear up space before adding
+        if (_cacheSize + ref->size() > _maxSize)
+            FreeMem(_maxSize);
```

Comment:
This reads like it Frees all cache. But FreeMem doesn't take an amount, it takes a threshold, so perhaps it doesn't. I am finding this line hard to read.

Reply:
This function may be also used to resize the cache to certain limit in case cache size setting changes. But indeed the argument should take ref->size() into account to, so it should be `_maxSize - ref->size()`.

I may rename FreeMem to something else to emphasize that it frees not for the exact amount but to keep at certain limit.